Added support for CNAME records add/remove

marank opened the pull request:
Signed-off-by: Matthias rank <development@m-rank.de>

**By submitting this pull request, I confirm the following:** `{please fill any appropriate checkboxes, e.g: [X]}`

`{Please ensure that your pull request is for the 'devel' branch!}`

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
- [X] I have Signed Off all commits. ([`git commit --signoff`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff))

---

**What does this PR aim to accomplish?:**
`{A detailed description, screenshots (if necessary), as well as links to any relevant GitHub issues}`
This PR enables adding/removing CNAME records via the admin interface.
To do so, the following PR needs to be accepted as well, because it allows editing the neccessary file using the pihole CLI: pi-hole/pi-hole#3351

**How does this PR accomplish the above?:**
`{A detailed description (such as a changelog) and screenshots (if necessary) of the implemented fix}`
![grafik](https://user-images.githubusercontent.com/7427683/81742024-bb38e380-949f-11ea-9ec5-5567a54ceb36.png)
I changed the menu accordingly to have both Local DNS records and CNAME records in one submenu.

**What documentation changes (if any) are needed to support this PR?:**
`{A detailed list of any necessary changes}`

marank committed:
Added support for CNAME records add/remove

Signed-off-by: Matthias rank <development@m-rank.de>

marank committed:
Removed already defined functions

Signed-off-by: Matthias Rank <development@m-rank.de>

DL6ER reviewed (dismissed):
I was thinking for a moment we should maybe fuse the DNS records and the CNAME records pages into one, however, it's likely clearer structured (esp. for the inexperienced) like you proposed it.

Some more thoughts:

- I merged the core PR into `development` already.
- There is a small merge conflict on this branch.

Yes, please move the routines over to `func.php` as you suggested. Just like I did in https://github.com/pi-hole/AdminLTE/pull/1276 and add support for importing CNAMEs through teleporter. There should be anything you need in #1276 where I did this for the local DNS records page.

Also, please add authentication checking for the CNAME page to avoid anyone being able to submit configuration even without logging into the web interface. See https://github.com/pi-hole/AdminLTE/commit/7b2d396295b41022ab9ed57ed00bca973b3c02ff for reference.

marank commented:
I'll update this PR tomorrow. Thanks for your review!

marank committed:
Added token verification
Moved CNAME functions to func.php
Added teleporter support
Fixed CNAME and DNS file declaration (filename was declarated in seperate file, therefore trying to read those files from func.php failed)
Fixed error and success response functions in func.php (calling addCustomDNSEntry and addCustomCNAMEEntry repeatedly from teleporter failed because of error function defined inside of those functions)

Signed-off-by: Matthias Rank <development@m-rank.de>

marank commented:
Okay, there were some hurdles while implementing the requested changes:

- I had to remove the definition of the `error` function inside of `AddCustomDNSEntry` because subsequent calls from teleporter failed (Cannot redeclare error() (previously declared in /var/www/html/admin/scripts/pi-hole/php/func.php:126) in /var/www/html/admin/scripts/pi-hole/php/func.php on line 126
- I had to move the declaration of the `$customDNSFile`/`$customCNAMEFile` vars to func.php because they didn't exist when called from teleporter
- When importing Local DNS Entries or Local CNAME Entries with teleporter and Clear existing data is set, every entry gets deleted seperately, always causing a restart of FTL (see webpage.sh). This has to be changed because FTL won't start if there were too many start attempts:
```
May 21 14:38:44 gradius systemd[1]: pihole-FTL.service: Start request repeated too quickly.
May 21 14:38:44 gradius systemd[1]: pihole-FTL.service: Failed with result 'start-limit-hit'.
```

I haven't used teleporter in the past, but all the tests I did were successfull.
On a side note, I'm not very familiar with resolving merge conflicts using git. I'd appreciate to get some help with that.

DL6ER commented:
@marank Sorry for the delay on this. You can use the internal conflict resolution mechanism of Github. You should see it above the tests in this PR:


![Screenshot at 2020-06-14 20-43-24](https://user-images.githubusercontent.com/16748619/84601369-dcfeff00-ae7f-11ea-9ddc-b4cb2c97cb6d.png)

It will open a page showing what was added to the modified pages meanwhile so you can review the changes and decide what is the best when fusing your new changes with the ones made meanwhile, for instance:
![Screenshot at 2020-06-14 20-45-53](https://user-images.githubusercontent.com/16748619/84601398-0ddf3400-ae80-11ea-85f8-824db24d7e1c.png)

marank committed:
Merge branch 'devel' into devel

marank commented:
Ok, I'll need some time to update this PR. There were some changes in the devel branch in the meantime which I want to add.

marank committed:
Incorporate upstream changes

Signed-off-by: Matthias Rank <development@m-rank.de>

marank commented:
I think I'm done, looks good to me. The only issue remaining is the one with FTL restarting too often when restoring a backup via teleporter. Maybe we should add a flag to the pihole command which prevents restarting FTL after every relevant add/remove DNS/CNAME command?

PromoFaux commented:
@DL6ER shall we change this to point at 5.1, or leave it til vNext?

DL6ER commented:
Merge now for the next version after v5.1 gives us more time to test things properly.

XhmikosR commented on scripts/pi-hole/js/customcname.js:
```diff
@@ -0,0 +1,123 @@
+var table;
+var token = $("#token").text();
+
+function showAlert(type, message) {
```
Why duplicate stuff here?

XhmikosR commented on scripts/pi-hole/js/customcname.js:
```diff
@@ -0,0 +1,123 @@
+
+/* global utils:false */
+
+var table;
```
Please avoid global variables

mcnahum commented:
I'm really waiting for that change to land.. any idea for when it is?

dschaper commented:
When it's ready. Which is the same as "soon". 

Addressing the issues that were brought up in https://github.com/pi-hole/AdminLTE/pull/1278#issuecomment-655138561 would help.
